### ISSUE TITLE
Support alternative calendars in telegram

### ISSUE THREAD

**Issue by afarsi** [telegramdesktop/tdesktop#4182]
In many countres, official calendar isn't Geoiregian calendar. For exmple, Arabic countries are using Hijri calendar as official celndar. And in many Asian countiries, people are using chinese or korean calendar. 
It will be a large project named alternative calendar and implement all of important alternative calenders in telegram.
For example, a user in arabic countries should be able to select Hijri calndar as preferd calndar(In telegram settings and it can be detected via system settings). And if he selcted it, he sould be able to see all of dates in Hjri date system, not georgian. For example 1407 Ramadn 05 not 2017 Jan 08. In addition, he should be able to search between two Hijri calendar dates not two georegian calendar dates.

Pay attention: It would be a huge and important benefit for telgram, because currently there is no messenger that supporting aletrnative calendars. 

You should support all of these calendars at least:
1-Persian Calendar =Iranian Calendar = Solar Hijri Calendar = Fix bug #1155 
2-Arabic Calendar = Lunar Hijri Calendar = Hijri Calendar
3-Chinese Calendar
4-Hebrew Calendar
5-Korean Calendar
Note1: You can read about all of these calendars in wikoedia and find covertation algorithms between these georgian dates and dates in these calendars.
Note2: Currently, google calendar, windows 10, Mac OS and gmail are supporting alternative calendars(For example, persian calendar)
Note3: Maybe icu can help you: http://userguide.icu-project.org/datetime/calendar

**Comment by neesti** [comment]
What's up from this?

**Comment by Aokromes** [comment]
imho still needed

**Comment by DearRude** [comment]
No updates on this?

**Comment by ilya-fedin** [comment]
> No updates on this?

Someone should design and create widgets for this

**Comment by neesti** [comment]
@afarsi 

 &rlm; چنانچه می‌دانید، نسخهٔ دسکتاپِ آی‌گرام تقویم جلالی هم دارد و سورس این برنامه هم در گیت‌هاب هست. آیا نمی‌شود که از آن برای افزودن تقویم جلالی به تگرامِ دسکتاپ یاری جست؟ &lrm;

**Comment by ilius** [comment]
Here is a Golang library I wrote, in case it helps someone
https://github.com/ilius/libgostarcal/tree/master/cal_types

**Comment by neesti** [comment]
Iranian calendar codes were added to the latest version of the Qt library:

https://codereview.qt-project.org/c/qt/qtbase/+/182341

**Comment by ilya-fedin** [comment]
> Iranian calendar codes were added to the latest version of the Qt library:

tdesktop doesn't use QCalendar for its calendar widget: https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/SourceFiles/ui/boxes/calendar_box.cpp

**Comment by atremiami** [comment]
Hello
How can I change the computer's Telegram calendar to the Hijri calendar?

**Comment by ilya-fedin** [comment]
You're asking in the issue which asks for this feature which means the feature doesn't exist, hence you just can't

**Comment by 23rd** [comment]
Does anybody want to test drive a build with Hijri and Solar Hijri calendars?
This is just an **experimental build** to try to collect some feedback and see if it would be good to leave this implementation  for future releases.

<details>
<summary>Screenshots</summary>

![image](https://github.com/user-attachments/assets/80de9d8e-112c-4d4e-81b2-137e8820235d) ![image](https://github.com/user-attachments/assets/a5b9882e-01c0-4116-86e8-a348638d94fd)

</details>

https://github.com/telegramdesktop/tdesktop/actions/runs/12828916288

**Comment by ilius** [comment]
I tested the [Linux build](https://github.com/telegramdesktop/tdesktop/actions/runs/12828916287/job/35773941984).
It looks good.

I'm not sure what's the "correct" name in English, but "Solar Hijri Calendar" is not a common term at all in English. "Persian Calendar" is a much more common and familiar name (and "Persian New Year", not "Solar Hijri New Year").

Also it would be nice to be able to change which day weeks start with (here starts with Saturday), since it's different than Sunday in so many countries.

**Comment by ilius** [comment]
I searched Google Books and the oldest usage of "Solar Hijri Calendar" I found was [a Pakistani magazine from 1965](https://www.google.com/books/edition/Illustrated_Weekly_of_Pakistan/tn06AQAAIAAJ?hl=en&gbpv=0&bsq=%22Solar%20Hijri%20Calendar%22)

![Image](https://github.com/user-attachments/assets/a0e6c662-d5d8-43ed-aabe-d3ee6cf5b359)

But try searching "Persian Calendar" and it gives so many results, even in 1700s, [here is one going back to 1738](https://www.google.com/books/edition/Cyclopaedia_Or_An_Universal_Dictionary_O/bKtqgJw9DiAC?hl=en&gbpv=1&bsq=%22Persian%20Calendar%22) (the term is much older though)

BTW, Hijri is pronounced Hejri in Persian, so it sounds/reads strange to Persian natives as well as English natives.

**Comment by afshin-ir** [comment]
Hello. 

I’d like to suggest revisiting this issue now that Telegram Desktop uses Qt 6.

Qt 6 has built-in support for the Persian (Jalali / Solar Hijri) calendar through `QCalendar::System::Jalali`. This means Telegram Desktop can potentially add Jalali calendar support without introducing a separate calendar-conversion library or maintaining a custom Jalali implementation.

A possible implementation would be:

* Add a **Calendar system** preference, with at least:

  * Gregorian
  * Persian (Jalali)
* When Persian is selected, use `QCalendar::System::Jalali` in the date-picker UI.
* Display Persian month names and Jalali years (e.g. فروردین، اردیبهشت، …، اسفند / ۱۴۰۵).
* Use Saturday as the first day of the week for the Persian calendar.
* Keep Telegram's internal date representation unchanged (`QDate` / absolute dates), and perform the calendar conversion only at the UI/input layer.
* Apply the selected calendar consistently to date-selection interfaces and, where appropriate, date-range/search interfaces.
* Avoid changing the dates transmitted through MTProto; this would only affect how dates are entered and displayed locally.

This seems particularly attractive because `QDate` already represents an absolute date independently of the calendar system, while `QCalendar` can provide the appropriate calendar-specific representation.

In other words, the implementation does not need to replace Telegram's existing Gregorian date handling. It could simply introduce a calendar abstraction at the UI level and use Qt's native Jalali calendar when the user selects Persian.

When this issue was originally opened, using ICU was discussed. With current Qt versions, however, the necessary Jalali calendar support is already available in Qt itself.

This could also provide a foundation for supporting other calendar systems in the future, rather than implementing Persian-specific date handling throughout the codebase.

I’d be very happy to see this reconsidered given the current Qt 6 capabilities.

**Comment by afshin-ir** [comment]
I took a closer look at the current date-picker implementation, and it seems that adding Jalali support could be relatively well-contained.

`CalendarBox` already receives and returns dates as `QDate`, while `ChooseDateTimeBox` keeps the selected date as a `QDate` and eventually converts it to a `QDateTime`/Unix timestamp. This suggests that the internal date representation and MTProto timestamps would not need to change.

The main places that appear to assume the Gregorian calendar are within `CalendarBox`, particularly:

* calculating the number of days in a month;
* calculating the number of rows in the month grid;
* converting a grid index to a date;
* moving between months;
* the month/year picker;
* generating the displayed month/year title.

Some of these operations currently rely directly on `QDate`'s Gregorian `year()`, `month()`, `daysInMonth()`, and `QDate(year, month, day)`, and the month/year picker also assumes 12 months per year.

These could potentially be moved to a `QCalendar` instance, for example:

```cpp
QCalendar _calendar;
```

with Gregorian as the default and `QCalendar::System::Jalali` when Persian is selected.

The relevant conversions could then use Qt's calendar API:

```cpp
_calendar.partsFromDate(date);
_calendar.dateFromParts(year, month, day);
_calendar.daysInMonth(month, year);
```

This would allow the UI to operate in Jalali years/months/days while continuing to use `QDate` internally.

The month/year picker would need the same treatment; otherwise the day grid could be Jalali while the month/year selector remained Gregorian.

The existing handling of the first day of the week also appears to be localized through `QLocale`, so the Persian locale could potentially provide the appropriate Saturday-based week layout without introducing a separate week calculation system.

For an initial implementation, I think it would be preferable to limit the scope to the date-picker/calendar UI rather than immediately changing every date displayed throughout Telegram. Message-date formatting and date-range search could then be addressed separately.

So, based on the current code, a possible minimal architecture would be:

`CalendarBox → QCalendar → Gregorian / Jalali`

while keeping:

`QDate → QDateTime → Unix timestamp`

unchanged.

This seems like a reasonably small and low-risk starting point for implementing the feature requested in this issue.

**Comment by ilya-fedin** [comment]
> I’d like to suggest revisiting this issue now that Telegram Desktop uses Qt 6.

No, Windows builds are still using Qt 5

**Comment by afshin-ir** [comment]
Thanks, that’s an important clarification.

In that case, perhaps the better approach would be to keep the calendar logic independent of Qt’s calendar APIs altogether, rather than making the feature depend on `QCalendar` (Qt 6).

A small calendar abstraction in Telegram Desktop could handle conversions between calendar-specific year/month/day values and the existing `QDate`, with Gregorian and Jalali as implementations. `CalendarBox` could then use that abstraction without caring which calendar system is selected.

That would keep the feature compatible with both the current Qt 5 Windows builds and Qt 6, while also providing a foundation for other alternative calendars mentioned in this issue.

I’ll look into what a minimal implementation of such an abstraction would look like in the current codebase.

**Comment by ilya-fedin** [comment]
iirc @23rd worked on some alternative calendar but idk whether it reached any release

**Comment by afshin-ir** [comment]
Hi @23rd — @ilya-fedin mentioned that you may have worked on an alternative calendar implementation at some point.

Do you happen to remember where that work was (PR, branch, or commit), and whether it was ever merged or released?

I’m looking into #4182 at the moment, so I’d be very interested to see what you had implemented back then.

**Comment by afshin-ir** [comment]
@ilya-fedin 
Another possibility: since Qt 6 already provides `QCalendar::System::Jalali`, would it make sense to initially enable the Jalali calendar only in Qt 6 builds?

That could allow the feature to be implemented and tested without requiring the Windows Qt 5 builds to support it immediately. Then, once the Windows builds eventually move to Qt 6, the same implementation could become available there as well.

This might be a reasonable incremental approach, while keeping the option of a Qt-independent calendar abstraction later if support for alternative calendars is needed in Qt 5 builds too.

**Comment by ilya-fedin** [comment]
> Another possibility: since Qt 6 already provides `QCalendar::System::Jalali`, would it make sense to initially enable the Jalali calendar only in Qt 6 builds?

I think you should discuss this with other devs

**Comment by afshin-ir** [comment]
@ilya-fedin 
Sure. Do you have any particular developers in mind that I should discuss this with?

**Comment by ilya-fedin** [comment]
@john-preston and @23rd

**Comment by afshin-ir** [comment]
Thanks. @john-preston @23rd — I was pointed to both of you regarding #4182.

@23rd, in particular, I’d be interested to know if you remember the alternative-calendar work mentioned above, and whether there’s any old PR, branch, or code that I could look at.

And @john-preston, I’d also be interested in your thoughts on the incremental Qt 6 approach I mentioned above, given that Qt 6 already provides `QCalendar::System::Jalali`.
